lib/tracecontext/tracestate.js

```diff
@@ -0,0 +1,309 @@
+
+        const indexOfValueLength = indexOfKey + lengthKey
+        const indexOfValue = indexOfValueLength + 1
+        const lengthValue = this.buffer[indexOfValueLength]
+
+        const key = this.buffer.slice(indexOfKey, indexOfKey + lengthKey).toString()
+        // bail out if this is our mutable namespace
+        if (key === this.listMemberNamespace) { continue }
+
+        // if this is not our key copy from the `0` byte to the end of the value
+        this.buffer.copy(newBuffer, newBufferOffset, i, indexOfValue + lengthValue)
+        newBufferOffset += (indexOfValue + lengthValue)
+
+        // skip ahead to first byte after end of value
+        i = indexOfValue + lengthValue - 1
+        continue
+      }
+    }
+
+    // now serialize the internal representation
+    const ourBytes = []
+    if (Object.keys(this.values).length > 0) {
+      // the zero byte
+      ourBytes.push(0)
+
+      // the length of the vendor namespace
+      ourBytes.push(this.listMemberNamespace.length)
+      // the chars of the vendor namespace
+      for (let i = 0; i < this.listMemberNamespace.length; i++) {
+        ourBytes.push(this.listMemberNamespace.charCodeAt(i))
+      }
+
+      // add the length of the value
+      const serializedValue = this._serializeValues(this.values)
+      ourBytes.push(serializedValue.length)
+
+      // add the bytes of the value
+      for (let i = 0; i < serializedValue.length; i++) {
+        ourBytes.push(serializedValue.charCodeAt(i))
+      }
+    }
+    const ourBuffer = Buffer.from(ourBytes)
+    return Buffer.concat(
+      [newBuffer, ourBuffer],
+      newBuffer.length + ourBuffer.length
+    ).toString('hex')
+  }
+
+  /**
+   * Returns JSON reprenstation of tracestate key/value pairs
+   *
+   * Does not parse the mutable list namespace
+   */
+  toJson () {
+    const json = {}
+
+    for (let i = 0; i < this.buffer.length; i++) {
+      const byte = this.buffer[i]
+      if (byte === 0) {
+        const indexOfKeyLength = i + 1
+        const indexOfKey = i + 2
+        const lengthKey = this.buffer[indexOfKeyLength]
+
+        const indexOfValueLength = indexOfKey + lengthKey
+        const indexOfValue = indexOfValueLength + 1
+        const lengthValue = this.buffer[indexOfValueLength]
+
+        const key = this.buffer.slice(indexOfKey, indexOfKey + lengthKey).toString()
+        const value = this.buffer.slice(indexOfValue, indexOfValue + lengthValue).toString()
+
+        json[key] = value
+
+        // skip ahead
+        i = indexOfValue + lengthValue - 1
+        continue
+      }
+    }
+
+    // then, serialize values from the internal representation. This means
+    // we end up prefering values set in this.value over
+    // values set in the initial buffer
+    json[this.listMemberNamespace] = this._serializeValues(
+      this.values
+    )
+    return json
+  }
+
+  toString () {
+    return this.toW3cString()
+  }
+
+  toW3cString () {
```

Comment:
The w3c spec states that the order of key/value pairs need to be preserved and that the modified keys need to be moved to the start:

> Vendors receiving a tracestate request header MUST send it to outgoing requests. It MAY mutate the value of this header before passing to outgoing requests. When mutating tracestate, the order of unmodified key/value pairs MUST be preserved. Modified keys MUST be moved to the beginning (left) of the list.

I think using `toJson` here might loose that order, since the `Object.entries` doesn't guarantee the order of the keys.

Reply:
Excellent catch -- I had not registered that portion of the specification. 

Also, in addition to the problem of plain objects not having a guaranteed order, this made me realize that we wern't moving our mutated `es` list-member to the left to the left side of the new `tracestate` header AND that we weren't maintaining the order of an unmutated `es` header. 

I've introduced a `toMap` method that ensure that the original order of the `tracestate` list-members is maintained, and added some new logic to ensure we don't remove `es` from the raw buffer until the first value is set.  This should ensure we conform to the letter of the W3C spec. There are also new tests to confirm the order is maintained/changed as specified.